tests/xgboost/test_xgboost_model_export.py

```diff
@@ -324,3 +324,34 @@ def test_sagemaker_docker_model_scoring_with_default_conda_env(xgb_model, model_
         check_dtype=False,
         check_less_precise=6,
     )
+
+
+# For now, run tests using this file
+def _canonicalize_version(version):
+    """
+    Canonicalizes a package version (e.g. '1.5.0-dev0' -> '1.5.0.dev0')
+    """
+    from packaging.version import Version
+
+    return str(Version(version))
+
+
+@pytest.mark.large
+def test_infer_pip_requirements(capsys):
+    import sklearn
+
+    model = xgb.XGBClassifier()
+
+    with mlflow.start_run():
+        mlflow.sklearn.log_model(model, artifact_path="model")
+        model_uri = mlflow.get_artifact_uri("model")
+
+    inferred_pip_reqs = infer_pip_requirements(model_uri)
+    assert f"xgboost=={_canonicalize_version(xgb.__version__)}" in inferred_pip_reqs
+    assert f"scikit-learn=={_canonicalize_version(sklearn.__version__)}" in inferred_pip_reqs
+
+    # DEBUG
+    with capsys.disabled():
+        print("\n" + "+" * 30)
+        print("\n".join(inferred_pip_reqs))
+        print("+" * 30)
```

Comment:
These lines print out the following in the `flavors` check **where lots of ML packages are installed**:

https://github.com/mlflow/mlflow/pull/4518/checks?check_run_id=2994199829#step:6:899

```diff
++++++++++++++++++++++++++++++
Bottleneck==1.3.2
PyYAML==5.4.1
Pygments==2.9.0
certifi==2021.5.30
chardet==4.0.0
click==7.1.2
cloudpickle==1.6.0
colorama==0.4.4
contextvars==2.4
databricks-cli==0.14.3
decorator==4.4.2
google-api-core==1.30.0
google-auth==1.32.1
google-cloud-core==1.7.1
google-cloud-storage==1.40.0
google-resumable-media==1.3.1
googleapis-common-protos==1.53.0
idna==2.10
immutables==0.15
importlib-metadata==4.6.1
ipython-genutils==0.2.0
ipython==7.16.1
jedi==0.18.0
joblib==1.0.1
matplotlib==3.3.4
mlflow-test-plugin==0.0.1
mlflow==1.18.1.dev0
numpy==1.19.5
packaging==21.0
pandas==1.1.5
parso==0.8.2
pexpect==4.8.0
pickleshare==0.7.5
prompt-toolkit==3.0.19
protobuf==3.17.3
py4j==0.10.9
pyspark==3.1.2
python-dateutil==2.8.1
pytz==2021.1
requests==2.25.1
ruamel.yaml.clib==0.2.6
ruamel.yaml==0.15.89
scikit-learn==0.24.2
scipy==1.5.4
setuptools==52.0.0.post20210125
six==1.15.0
traitlets==4.3.3
typing-extensions==3.7.4.3
urllib3==1.26.6
wcwidth==0.2.5
xgboost==1.4.2
zipp==3.5.0
++++++++++++++++++++++++++++++
```

- `importlib_medata`

Reply:
One interesting behavior of `sys.modules` that I found is:

```diff
>>> import sys
>>> sys.modules.keys()
[
    "sys",
    "builtins",
    "_frozen_importlib",
    "_imp",
    "_thread",
    "_warnings",
    "_weakref",
    "zipimport",
    "_frozen_importlib_external",
    "_io",
    "marshal",
    "posix",
    "encodings",
    "codecs",
    "_codecs",
    "encodings.aliases",
    "encodings.utf_8",
    "_signal",
    "__main__",
    "encodings.latin_1",
    "io",
    "abc",
    "_abc",
    "site",
    "os",
    "stat",
    "_stat",
    "_collections_abc",
    "posixpath",
    "genericpath",
    "os.path",
    "_sitebuiltins",
    "_bootlocale",
    "_locale",
    "types",
    "importlib",
    "importlib._bootstrap",
    "importlib._bootstrap_external",
    "warnings",
    "importlib.util",
    "importlib.abc",
    "importlib.machinery",
    "contextlib",
    "collections",
    "operator",
    "_operator",
    "keyword",
    "heapq",
    "_heapq",
    "itertools",
    "reprlib",
    "_collections",
    "functools",
    "_functools",
+    "google",
+    "google.cloud",
+    "google.logging",
+    "mpl_toolkits",
    "ruamel",
+    "sphinxcontrib",
    "readline",
    "atexit",
    "rlcompleter",
]
```

Reply:
Looks like some namespace packages are loaded during the REPL startup.